File: 0383-ransom-note/0383-ransom-note.py
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        possible = True
        word_map = {}
        for ch in magazine:
            word_map[ch] = word_map.get(ch,0) + 1
        #magazine check
        for ch in ransomNote:
            if ch in word_map:
                word_map[ch] -= 1
                if word_map[ch] == 0:
                    del word_map[ch]
            else:
                possible = False
                break        
        return possible

    # def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        # ct1, ct2 = Counter(ransomNote), Counter(magazine)        
        # return True if ct1 & ct2 == ct1 else False

File: 0383-ransom-note/0383-ransom-note.java
class Solution {
    public boolean canConstruct(String ransomNote, String magazine) {
        //create hashmap
        Map<Character, Integer> wordCount = new HashMap<>();
        for(char ch : magazine.toCharArray()){
            wordCount.put(ch, wordCount.getOrDefault(ch,0)+1);
        }

        for(char ch : ransomNote.toCharArray()){
            if(wordCount.get(ch) != null){
                int count = wordCount.get(ch);
                count--;
                if(count == 0){
                    wordCount.remove(ch);
                }else{
                wordCount.put(ch, count);
                }
            }else{
                return false;
            }
        }
        return true;

    }
}